src/Network/OAuth/OAuth2/Internal.hs

```diff
@@ -76,63 +81,90 @@ type URI = BS.ByteString
                          -> (URI, PostBody)  -- ^ refresh token request URL plus the request body.
-refreshAccessTokenUrl oa rtoken = (uri, body)
+refreshAccessTokenUrl oa token = (uri, body)
   where uri = oauthAccessTokenEndpoint oa
-        body = transform' [ ("grant_type", Just "refresh_token")
-                          , ("refresh_token", Just rtoken)
-                          ]
-
---------------------------------------------------
--- * UTILs
---------------------------------------------------
-
--- | Append query parameters using `"?"` or `"&"`.
-appendQueryParam :: URI -> QueryParams -> URI
-appendQueryParam uri q = if "?" `BS.isInfixOf` uri
-                         then uri `BS.append` "&" `BS.append` renderSimpleQuery False q
-                         else uri `BS.append` renderSimpleQuery True q
+        body = [ ("grant_type", "refresh_token")
+               , ("refresh_token", encodeUtf8 $ rtoken token)
+               ]
 
 -- | For `GET` method API.
-appendAccessToken :: URI               -- ^ Base URI
-                     -> AccessToken    -- ^ Authorized Access Token
-                     -> URI            -- ^ Combined Result
-appendAccessToken uri t = appendQueryParam uri (accessTokenToParam t)
+appendAccessToken :: URIRef a             -- ^ Base URI
+                     -> AccessToken       -- ^ Authorized Access Token
+                     -> URIRef a          -- ^ Combined Result
+appendAccessToken uri t = over (queryL . queryPairsL) (\query -> query ++ (accessTokenToParam t)) uri
 
 -- | Create 'QueryParams' with given access token value.
-accessTokenToParam :: AccessToken -> QueryParams
-accessTokenToParam at = [("access_token", accessToken at)]
-
-
--- | Lift value in the 'Maybe' and abandon 'Nothing'.
-transform' :: [(a, Maybe b)] -> [(a, b)]
-transform' = map (\(a, Just b) -> (a, b)) . filter (isJust . snd)
+accessTokenToParam :: AccessToken -> [(BS.ByteString, BS.ByteString)]
+accessTokenToParam t = [("access_token", encodeUtf8 $ atoken t)]
+
+appendQueryParams :: [(BS.ByteString, BS.ByteString)] -> URIRef a -> URIRef a
+appendQueryParams params =
+  over (queryL . queryPairsL) (params ++ )
+
+uriToRequest :: MonadThrow m => URI -> m Request
+uriToRequest uri = do
+  ssl <- case (view (uriSchemeL . schemeBSL) uri) of
+    "http" -> return False
+    "https" -> return True
+    s -> throwM $ InvalidUrlException (show uri) ("Invalid scheme: " ++ show s)
+  let
+    query = fmap (\(a, b) -> (a, Just b)) (view (queryL . queryPairsL) uri)
+    hostL = (authorityL . _Just . authorityHostL . hostBSL)
+    portL = (authorityL . _Just . authorityPortL . _Just . portNumberL)
+    defaultPort = (if ssl then 443 else 80) :: Int
+
+    req = (setQueryString query) $ defaultRequest {
+        secure = ssl,
+        path = (view pathL uri)
+      }
+    req2 = (over hostLens . maybe id const . preview hostL) uri req
+    req3 = (over portLens . maybe (\_ -> defaultPort) const . preview portL) uri req2
+  return $ req3
```

Comment:
@reactormonk Do you remember why we add this method instead of 
https://hackage.haskell.org/package/http-client-0.7.18/docs/Network-HTTP-Client.html#v:parseRequest

Reply:
Yeet at will, it's been too long.

Reply:
I bet it's b/c those are differ URI, `URI.ByteString` v.s. `Network.URI`.

I'll try to investigate to see if it can be 

```
uriToRequest = requestRequest . show 
```

Reply:
Actually https://hackage.haskell.org/package/network-uri-2.6.4.2
does have `uri` QQ as well. 
Shall probably consider to switch over